com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -319,6 +324,25 @@ public void onClassRedefined(Consumer<List<String>> consumer) {
 
     @Override
     public CompletableFuture<List<String>> redefineClasses() {
+        try {
+            IProject mainProject = null;
+            List<IJavaProject> javaProjects = ResolveClasspathsHandler.getJavaProjectFromType(context.getMainClass());
```

Comment:
This is what I can find to target the main project for the session. It might be a little bit expensive.

Reply:
Suggest to move the new logic to a separate function like triggerBuildIfBspProject.

Check if the workspace contains any bsp project firstly, and skip it directly if there are no bsp projects.
And fetch the project name from options (`(String) options.get(Constants.PROJECT_NAME)`) , if empty, then get it from main class. The `options` is passed when initializing hotCodeReplaceProvider, you can save it in the context.

https://github.com/microsoft/java-debug/blob/83403a458e0d01fba5a3871fea81af742460bdf0/com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java#L265